node/core/prospective-parachains/src/fragment_tree.rs

```diff
@@ -0,0 +1,629 @@
+				if self.nodes.last().map_or(true, |last| last.parent == NodePointer::Root) {
+					self.nodes.push(node);
+				} else {
+					let pos =
+						self.nodes.iter().take_while(|n| n.parent == NodePointer::Root).count();
+					self.nodes.insert(pos, node);
+				}
+			},
+		}
+	}
+
+	fn node_has_candidate_child(
+		&self,
+		pointer: NodePointer,
+		candidate_hash: &CandidateHash,
+	) -> bool {
+		self.node_candidate_child(pointer, candidate_hash).is_some()
+	}
+
+	fn node_candidate_child(
+		&self,
+		pointer: NodePointer,
+		candidate_hash: &CandidateHash,
+	) -> Option<NodePointer> {
+		match pointer {
+			NodePointer::Root => self
+				.nodes
+				.iter()
+				.take_while(|n| n.parent == NodePointer::Root)
+				.enumerate()
+				.find(|(_, n)| &n.candidate_hash == candidate_hash)
+				.map(|(i, _)| NodePointer::Storage(i)),
+			NodePointer::Storage(ptr) =>
+				self.nodes.get(ptr).and_then(|n| n.candidate_child(candidate_hash)),
+		}
+	}
+
+	/// Returns a set of candidate hashes contained in nodes.
+	///
+	/// This runs in O(n) time in the number of nodes
+	/// and allocates memory.
+	pub(crate) fn candidates(&self) -> HashSet<CandidateHash> {
+		let mut set = HashSet::with_capacity(self.nodes.len());
+
+		for f in &self.nodes {
+			set.insert(f.candidate_hash);
+		}
+
+		set
+	}
+
+	/// Add a candidate and recursively populate from storage.
+	pub(crate) fn add_and_populate(&mut self, hash: CandidateHash, storage: &CandidateStorage) {
+		let candidate_entry = match storage.get(&hash) {
+			None => return,
+			Some(e) => e,
+		};
+
+		let candidate_parent = &candidate_entry.candidate.persisted_validation_data.parent_head;
+
+		// Select an initial set of bases, whose required relay-parent matches that of the candidate.
+		let root_base = if &self.scope.base_constraints.required_parent == candidate_parent {
+			Some(NodePointer::Root)
+		} else {
+			None
+		};
+
+		let non_root_bases = self
+			.nodes
+			.iter()
+			.enumerate()
+			.filter(|(_, n)| {
+				n.cumulative_modifications.required_parent.as_ref() == Some(candidate_parent)
+			})
+			.map(|(i, _)| NodePointer::Storage(i));
+
+		let bases = root_base.into_iter().chain(non_root_bases).collect();
+
+		// Pass this into the population function, which will sanity-check stuff like depth, fragments,
+		// etc. and then recursively populate.
+		self.populate_from_bases(storage, bases);
+	}
+
+	/// Select a candidate after the given `required_path` which pass
+	/// the predicate.
+	///
+	/// If there are multiple possibilities, this will select the first one.
+	///
+	/// This returns `None` if there is no candidate meeting those criteria.
+	///
+	/// The intention of the `required_path` is to allow queries on the basis of
+	/// one or more candidates which were previously pending availability becoming
+	/// available and opening up more room on the core.
+	pub(crate) fn select_child(
+		&self,
+		required_path: &[CandidateHash],
```

Comment:
I am a bit confused by this `required_path`. Given that `FragmentTree`s are already per leaf, I would assume we already know what candidates for that para have been backed on that chain already and thus that this tree is already based on that root.

Reply:
`required_path` is like an `OccupiedCoreAssumption`. `para_head(relay_parent, para_id)` is the root of the tree, but it doesn't account for the candidate pending availability. This does.

Reply:
This is all documented.

Reply:
> This is all documented.

I know, but I could not make sense of it at the time, but I think I might now. Thanks for the pointers!